backend/main/server/handlers/communities.go

```diff
@@ -0,0 +1,238 @@
+			return
+		}
+		c.Cid = &pin.IpfsHash
+
+		validate := validator.New()
+		vErr := validate.Struct(c)
+		if vErr != nil {
+			log.Error().Err(vErr).Msg("Invalid community.")
+			s.RespondWithError(w, http.StatusBadRequest, vErr.Error())
+			return
+		}
+
+		// Create Community
+		if err := c.CreateCommunity(db); err != nil {
+			log.Error().Err(err).Msg("Database error creating community.")
+			s.RespondWithError(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+		// Grant roles to community creator
+		if err := models.GrantRolesToCommunityCreator(db, c.Creator_addr, c.ID); err != nil {
+			log.Error().Err(err).Msg("Database error adding community creator roles.")
+			s.RespondWithError(w, http.StatusInternalServerError, err.Error())
+		}
+
+		// Grant any additional roles
+		if payload.Additional_admins != nil {
+			for _, addr := range *payload.Additional_admins {
+				if err := models.GrantAdminRolesToAddress(db, c.ID, addr); err != nil {
+					s.RespondWithError(w, http.StatusInternalServerError, err.Error())
+					return
+				}
+			}
+		}
+
+		if payload.Additional_authors != nil {
+			for _, addr := range *payload.Additional_authors {
+				if err := models.GrantAuthorRolesToAddress(db, c.ID, addr); err != nil {
+					s.RespondWithError(w, http.StatusInternalServerError, err.Error())
+					return
+				}
+			}
+		}
+
+		s.RespondWithJSON(w, http.StatusCreated, c)
+	}
+}
+
+func updateCommunity(db *s.Database, fa *s.FlowAdapter, config s.Config) func (w http.ResponseWriter, r *http.Request) {
+	return func (w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		id, err := strconv.Atoi(vars["id"])
+		if err != nil {
+			s.RespondWithError(w, http.StatusBadRequest, "Invalid community ID.")
+			return
+		}
+		var payload models.UpdateCommunityRequestPayload
+
+		decoder := json.NewDecoder(r.Body)
+		if err := decoder.Decode(&payload); err != nil {
+			log.Error().Err(err)
+			s.RespondWithError(w, http.StatusBadRequest, "Invalid request payload.")
+			return
+		}
+
+		defer r.Body.Close()
+
+		var c = models.Community{ID: id}
+		if err := c.GetCommunity(db); err != nil {
+			log.Error().Err(err)
+			s.RespondWithError(w, http.StatusBadRequest, fmt.Sprintf("Invalid request: no community with ID %d.", id))
+			return
+		}
+
+		// validate is commuity creator
+		// TODO: update to validating address is admin
+		if err := c.CanUpdateCommunity(db, payload.Signing_addr); err != nil {
+			log.Error().Err(err)
+			s.RespondWithError(w, http.StatusForbidden, err.Error())
+			return
+		}
+
+		// validate timestamp of request/message
+		if err := validateTimestamp(payload.Timestamp, 60, config); err != nil {
+			log.Error().Err(err)
+			s.RespondWithError(w, http.StatusForbidden, err.Error())
+			return
+		}
+
+		// validate user signature
+		if err := validateSignature(payload.Signing_addr, payload.Timestamp, payload.Composite_signatures, fa, config); err != nil {
+			s.RespondWithError(w, http.StatusForbidden, err.Error())
```

Comment:
group all the valudation functions into one function something like, `ValidateUser`?

Reply:
Done. Also created a helper function that can validate user role as well, since I saw several places where that logic. was duplicated.